Refactor Index Rebuild

KatrinIhler opened the pull request:
This removes classes from the index service module and puts them into two new modules:

- SearchIndex: The AbstractSearchIndex and all directly related classes.
- IndexRebuild: All classes that send or receive messages for the IndexRebuild.

Additionally to classes from the IndexService, these modules picked up some other related classes that were scattered in the rest of the code base for some reason. The actual changes aren't all that much, most of this PR consists of dependency and import adjustments.

Important: This does not improve the IndexRebuild itself in any way, but was a way for me to get started and get a better overview over what does what, and is imo an improvement to before, especially because it separates the AbstractSearchIndex, which is quite reasonable, from the madness that is the IndexService.

I wanted to get this merged ahead of other future changes because conflicts with new changes can arise quickly and they will be a hassle to fix.

Edit: The IndexRebuild module might be refactored further in the future, but as I wasn't quite sure what was necessary yet, I'll leave it as is for now.

KatrinIhler committed:
Move Index Rebuild to its own service

Remove the index rebuild from the AbstractSearchIndex and put it into
it's own service and module to simplify dependencies.

KatrinIhler committed:
Move AbstractSearchIndex to new module

Move the AbstractSearchIndex and related classes to its own module so
other modules can use it without depending on the index service.

KatrinIhler committed:
Rename IndexRebuild module, add IndexProducer

Rename IndexRebuildService module to IndexRebuild, add IndexProducer
classes to keep all parts related to the index rebuild mesages in one
spot.

karendolan commented:
Hi @KatrinIhler, is there overarching documentation about the Opencast service indices? For example, how the admin search indices are related to the other service indices?

KatrinIhler commented:
@karendolan As far as I know there is no overarching documentation. This PR affects the two ElasticSearch Indexes for the Admin UI and the External API. This has nothing to do with the Search Endpoint and its index, or other indexes e.g. for the WorkflowService. 
For more clarity, I'm considering renaming the relevant modules to contain _elasticsearch_ instead of _search_. But I don't know yet when I will get to this.

KatrinIhler committed:
Fix: Re-add Comments Index

LukasKalbertodt committed:
Fix checkstyle violations for `index-rebuild` and `search-index` modules

LukasKalbertodt committed:
Remove useless doc comment links

These attempted to link to the abstract method that was overwritten
by the commented method. But:

- This is not necessary. Javadoc already detects, shows and links the
  overwritten method.
- The links were broken/outdated.
- The line was longer than 120 chars which is why it was noticed by
  checkstyle.

lkiesow commented on modules/admin-ui-index/src/main/java/org/opencastproject/adminui/index/AdminUISearchIndex.java:
```diff
@@ -22,13 +22,13 @@
-import org.opencastproject.index.service.impl.index.series.Series;
-import org.opencastproject.index.service.impl.index.theme.Theme;
-import org.opencastproject.index.service.impl.index.theme.ThemeIndexSchema;
+import org.opencastproject.search.index.AbstractSearchIndex;
+import org.opencastproject.search.index.event.Event;
+import org.opencastproject.search.index.event.EventIndexSchema;
+import org.opencastproject.search.index.group.Group;
+import org.opencastproject.search.index.series.Series;
+import org.opencastproject.search.index.theme.Theme;
+import org.opencastproject.search.index.theme.ThemeIndexSchema;
```
These now use the same namespace as the completely unrelated search service.
That's confusing. Can we use something else?

https://github.com/opencast/opencast/blob/e9f79c2d585e3c5facc24d04681753b948704041/modules/search-service-api/src/main/java/org/opencastproject/search/api/SearchService.java#L23

KatrinIhler replied:
Good point. I'm gonna try to rename this into something else, probably "elasticsearch".

KatrinIhler replied:
I renamed this module and its namespace and also the modules search and search-api, which had the same problem, from `search` to `elasticsearch` (and added an `-impl` in one case). Things shouldn't be quite so confusing now. 

I didn't touch class names for now, otherwise I'd still be here next week. The same goes for modules with `index` in their name, since then I'd have to move the index service, and I don't want to deal with that right now. :D

KatrinIhler committed:
Rename module search-index

KatrinIhler committed:
Rename search and search-api modules

KatrinIhler committed:
Fix checkstyle errors

KatrinIhler committed:
Fix import order

KatrinIhler committed:
Fix OSGI wiring errors

KatrinIhler commented:
Checks fell over for some reason, but now everything is fine. :D

lkiesow commented:
I can take the review but it would be nice if you could talk me through the important parts of this patch in the next couple of days so that I do not have to read through all the 175 modified files.

KatrinIhler commented:
@lkiesow No problem! I'm gonna ping you so we can talk about a possible time. :D

pascalseeland reviewed (approved):
Looks good.

KatrinIhler commented:
Pascal made me aware that Intellij did some unintentional changes while fixing the import order. I'm gonna revert them, so don't merge just yet. :D

pascalseeland commented:
I did accept it even with these changes, since they are basically like code cleanup. I'm totally fine with reverting them, but I think should take a look into the entire test suite of the external api why it does need those files all together. As far as I know, it does only use json.

KatrinIhler committed:
Revert unintentional formatting changes

KatrinIhler commented:
@pascalseeland I reverted them (and similar ones) for now, since these changes have nothing to do with what I'm trying to accomplish here. Maybe we can throw some of them away, that wouldn't surprise me. But that's something for later. :)

pascalseeland commented:
I also approve the updated version.

KatrinIhler commented:
Since Pascal reviewed this (thanks for that btw!) and was satisfied, but cannot merge this, I'm gonna go ahead and do that now. :)
